xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/UserMentionEventListener.java

```diff
@@ -0,0 +1,200 @@
+    public void onEvent(Event event, Object source, Object data)
+    {
+        // We send notification to the mentions explicitly typed as local user, but also untyped mentions (i.e., with a
+        // null type).
+        MentionNotificationParameters mentionNotificationParameters = (MentionNotificationParameters) data;
+        Map<String, Set<MentionNotificationParameter>> newMentionsMap = mentionNotificationParameters.getNewMentions();
+        Set<MentionNotificationParameter> newMentions = new HashSet<>();
+        Set<MentionNotificationParameter> untypedMentionsMap = newMentionsMap.get(null);
+        if (untypedMentionsMap != null) {
+            newMentions.addAll(untypedMentionsMap);
+        }
+        Set<MentionNotificationParameter> userMentionsMap = newMentionsMap.get(TYPE);
+        if (userMentionsMap != null) {
+            newMentions.addAll(userMentionsMap);
+        }
+
+        Optional<XWikiDocument> optionalDoc;
+        EntityReference entityReference = mentionNotificationParameters.getEntityReference();
+        try {
+            optionalDoc = Optional.of(this.documentRevisionProvider
+                .getRevision((DocumentReference) entityReference.extractReference(EntityType.DOCUMENT),
+                    mentionNotificationParameters.getVersion()));
+        } catch (XWikiException e) {
+            this.logger.warn("Failed to send the mentions notifications. Cause: [{}]", getRootCauseMessage(e));
+            optionalDoc = Optional.empty();
+        }
+        optionalDoc.ifPresent(doc -> {
+            for (MentionNotificationParameter mentionNotificationParameter : newMentions) {
+                handleNotification(mentionNotificationParameters, entityReference, doc, mentionNotificationParameter);
+            }
+        });
+    }
+
+    private void handleNotification(MentionNotificationParameters mentionNotificationParameters,
+        EntityReference entityReference, XWikiDocument doc, MentionNotificationParameter mentionNotificationParameter)
+    {
+        String extractedQuote =
+            getQuoteFromXDOM(mentionNotificationParameters, entityReference, doc, mentionNotificationParameter);
+
+        MentionEventParams params = new MentionEventParams()
+            .setUserReference(mentionNotificationParameters.getAuthorReference())
+            .setDocumentReference(
+                this.entityReferenceSerializer.serialize(entityReference.extractReference(EntityType.DOCUMENT)))
+            .setLocation(mentionNotificationParameters.getLocation())
+            .setAnchor(mentionNotificationParameter.getAnchorId())
+            .setQuote(extractedQuote);
+        String mentionedIdentity = mentionNotificationParameter.getReference();
+
+        DocumentReference documentUserReference = this.documentReferenceResolver.resolve(mentionedIdentity,
+            entityReference.extractReference(EntityType.WIKI));
+        String identity = this.entityReferenceSerializer.serialize(documentUserReference);
```

Comment:
Why don't use a UserReferenceResolver and a UserReferenceSerializer here? a priori you're only handling user mentions there.

Reply:
You are right, user resolver/serializer used instead.